tests/test_pex_warnings.py

```diff
@@ -0,0 +1,76 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import warnings
+
+import pytest
+
+from pex import pex_warnings
+from pex.compatibility import PY2
+from pex.pex_info import PexInfo
+from pex.pex_warnings import PEXWarning
+from pex.variables import Variables
+
+
+def exercise_warnings(pex_info, **env):
+  with warnings.catch_warnings(record=True) as events:
+    warnings.resetwarnings()
+    pex_warnings.configure_warnings(pex_info, env=Variables(environ=env))
+    pex_warnings.warn('test')
+    return events
+
+
+def assert_warnings(pex_info, **env):
+  events = exercise_warnings(pex_info, **env)
+  assert 1 == len(events)
+  warning = events[0]
+  assert PEXWarning == warning.category
+  assert 'test' == str(warning.message)
+
+
+def assert_no_warnings(pex_info, **env):
+  events = exercise_warnings(pex_info, **env)
+  assert 0 == len(events)
+
+
+def pex_info_no_emit_warnings():
+  pex_info = PexInfo.default()
+  pex_info.emit_warnings = False
+  return pex_info
+
+
+skip_py2_warnings = pytest.mark.skipif(PY2,
+                                       reason="The warnings.catch_warnings mechanism doesn't work "
+                                              "properly under python 2.7 / pypy2, etc. across "
+                                              "multiple tests")
```

Comment:
It'd be helpful to specify why. Sounds like you investigated this and it has to do with C code. Any knowledge transfer would be helpful in case someone spends more time trying to figure this out.

Reply:
I have to disagree here. This will never be fixed in python 2.7 since it's eol and pypy appears to be trying to emulate CPython as much as possible - which makes sense. As such the detailed why is unimportant.